decision/engine.go

```diff
@@ -430,85 +509,248 @@ func (e *Engine) MessageReceived(ctx context.Context, p peer.ID, m bsmsg.BitSwap
+				Work:     entrySize,
+				Data: &taskData{
+					BlockSize:    blockSize,
+					HaveBlock:    true,
+					IsWantBlock:  isWantBlock,
+					SendDontHave: entry.SendDontHave,
+				},
+			})
 		}
 	}
+
+	// Push entries onto the request queue
 	if len(activeEntries) > 0 {
-		e.peerRequestQueue.PushBlock(p, activeEntries...)
+		e.peerRequestQueue.PushTasks(p, activeEntries...)
 	}
-	for _, block := range m.Blocks() {
-		log.Debugf("got block %s %d bytes", block, len(block.RawData()))
-		l.ReceivedBytes(len(block.RawData()))
+}
+
+// Get the size of a HAVE / HAVE_NOT entry
+func (e *Engine) getBlockPresenceSize(c cid.Cid) int {
+	return bsmsg.BlockPresenceSize(c)
+}
+
+// Split the want-have / want-block entries from the cancel entries
+func (e *Engine) splitWantsCancels(es []bsmsg.Entry) ([]bsmsg.Entry, []bsmsg.Entry) {
+	var wants []bsmsg.Entry
+	var cancels []bsmsg.Entry
+	for _, et := range es {
+		if et.Cancel {
+			cancels = append(cancels, et)
+		} else {
+			wants = append(wants, et)
+		}
 	}
+	return wants, cancels
 }
 
-func (e *Engine) addBlocks(ks []cid.Cid) {
-	work := false
+// ReceiveFrom is called when new blocks are received and added to the block
+// store, meaning there may be peers who want those blocks, so we should send
+// the blocks to them.
+func (e *Engine) ReceiveFrom(from peer.ID, blks []blocks.Block, haves []cid.Cid) {
+	// If we've received a block or a HAVE from a peer, we can remove any wants
+	// that peer sent us for the block
+	e.removeReceivedWants(from, blks, haves)
+
+	if len(blks) == 0 {
+		return
+	}
 
+	// Get the size of each block
+	blockSizes := make(map[cid.Cid]int)
+	for _, blk := range blks {
+		blockSizes[blk.Cid()] = len(blk.RawData())
+	}
+
+	// Check each peer to see if it wants one of the blocks we received
+	work := false
+	e.lock.RLock()
 	for _, l := range e.ledgerMap {
-		l.lk.Lock()
-		for _, k := range ks {
+		l.lk.RLock()
+
+		for _, b := range blks {
+			k := b.Cid()
+
 			if entry, ok := l.WantListContains(k); ok {
-				e.peerRequestQueue.PushBlock(l.Partner, peertask.Task{
-					Identifier: entry.Cid,
-					Priority:   entry.Priority,
-				})
 				work = true
+
+				blockSize := blockSizes[k]
+				isWantBlock := e.sendAsBlock(entry.WantType, blockSize)
+
+				// if isWantBlock {
+				// 	log.Debugf("  add-block put rq %s->%s %s as want-block (%d bytes)\n", lu.P(e.self), lu.P(l.Partner), lu.C(k), blockSize)
+				// } else {
+				// 	log.Debugf("  add-block put rq %s->%s %s as want-have (%d bytes)\n", lu.P(e.self), lu.P(l.Partner), lu.C(k), blockSize)
+				// }
+
+				entrySize := blockSize
+				if !isWantBlock {
+					entrySize = e.getBlockPresenceSize(k)
+				}
+
+				e.peerRequestQueue.PushTasks(l.Partner, peertask.Task{
+					Topic:    entry.Cid,
+					Priority: entry.Priority,
+					Work:     entrySize,
+					Data: &taskData{
+						BlockSize:    blockSize,
+						HaveBlock:    true,
+						IsWantBlock:  isWantBlock,
+						SendDontHave: false,
+					},
+				})
```

Comment:
This code is duplicated in https://github.com/ipfs/go-bitswap/pull/189/files#diff-cf2f06f9713472b2929ed047479de960R602-R628. We might want to factor this out into a `newBlockTask` function (maybe).

Reply:
It's basically all parameters, so I don't think putting it into a function will make it any more concise